File: leetcode/2265. Count Nodes Equal to Average of Subtree/main.py
import math

from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        self.counter = 0

        self.dfs(root)

        return self.counter

    def dfs(self, root: Optional[TreeNode]):
        if not root:
            return

        self.dfs(root.left)
        if root.val == self.find_node_avg(root):
            self.counter += 1
        self.dfs(root.right)

    def find_node_avg(self, root: Optional[TreeNode]) -> int:
        avg_node_vals = []
        self.inorder_traversal_to_find_avg(root, avg_node_vals)

        return math.floor(sum(avg_node_vals) / len(avg_node_vals))

    def inorder_traversal_to_find_avg(self, root: Optional[TreeNode], avg_node_vals: List[int]):
        if not root:
            return

        self.inorder_traversal_to_find_avg(root.left, avg_node_vals)
        avg_node_vals.append(root.val)
        self.inorder_traversal_to_find_avg(root.right, avg_node_vals)

File: leetcode/2265. Count Nodes Equal to Average of Subtree/main.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn average_of_subtree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut counter = 0;
        Solution::dfs(&root, &mut counter);
        counter
    }

    fn dfs(root: &Option<Rc<RefCell<TreeNode>>>, counter: &mut i32) {
        if let Some(node) = root {
            Solution::dfs(&node.borrow().left, counter);
            if node.borrow().val == Solution::find_average(root) {
                *counter += 1
            }
            Solution::dfs(&node.borrow().right, counter);
        }
    }

    fn find_average(root: &Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut avg_node_vals = vec![];
        Solution::inorder_traversal_to_find_average(root, &mut avg_node_vals);

        let mut sum = 0;
        for val in &avg_node_vals {
            sum += val
        }

        sum / avg_node_vals.len() as i32
    }

    fn inorder_traversal_to_find_average(
        root: &Option<Rc<RefCell<TreeNode>>>,
        avg_nodes: &mut Vec<i32>,
    ) {
        if let Some(node) = root {
            Solution::inorder_traversal_to_find_average(&node.borrow().left, avg_nodes);
            avg_nodes.push(node.borrow().val);
            Solution::inorder_traversal_to_find_average(&node.borrow().right, avg_nodes);
        }
    }
}
